### PR TITLE
Show diff report in accessible colors and reveal hidden line characters

### LINKED ISSUE #116: Syrupy assertion diff does not show missing carriage return
**Describe the bug**

Syrupy assertion diff does not show missing carriage return.

**To Reproduce**

Add this test.

```python
def test_example(snapshot):
    assert snapshot == "line 1\r\nline 2"
```

Run `pytest --snapshot-update`.

Remove the `\r` from the string in the test case so you get:

```python
def test_example(snapshot):
    assert snapshot == "line 1\nline 2"
```

Run `pytest`. The test will fail with a useless diff of "...".

**Expected behavior**

Syrupy should output each line with the missing carriage return, with some indication of what's missing.

**Additional context**

Syrupy v0.3.1

#113 fixed the bug where carriage returns were not being serialized. This issue addresses the missing functionality in the snapshot assertion diff reporter.

### LINKED ISSUE #120: Account for accessibility and readability in snapshot outputs
This is a general issue to track notes about readability and accessibility in diffs. See this jest blog post for more information: https://jestjs.io/blog/2020/01/21/jest-25, and this jest PR https://github.com/facebook/jest/pull/9132

They seem to have put a lot of effort into improving diff outputs recently so it's definitely something to consider and take inspiration from.

Sorry for the lack of specificity on this issue, mostly created it for tracking purposes with the idea of triggering an investigation.

### PR DESCRIPTION
**Pull request by noahnu**
## Description

Refactors the diff reporter to improve accessibility and show diff on newlines and carriage returns.

## Related Issues

- Closes #116 `\r\n`
- Closes #120 `a11y`

## Checklist

- [x] This PR has sufficient test coverage.
- [x] I have updated the CHANGELOG.md.

## Additional Comments

No additional comments.

### CODE REVIEW

File: src/syrupy/extensions/base.py [hunk 1]
```diff
@@ -218,31 +227,143 @@ def diff_lines(
         for line in self.__diff_lines(str(snapshot_data), str(serialized_data)):
             yield reset(line)
 
+    @property
+    def _ends(self) -> Dict[str, str]:
+        return {"\n": self._marker_new_line, "\r": self._marker_carriage}
+
+    @property
+    def _context_line_count(self) -> int:
+        return 1
```
Comment (iamogbz) [inline src/syrupy/extensions/base.py]: Previous implementation is equivalent to setting this to `0`

File: src/syrupy/extensions/amber.py [hunk 2]
```diff
@@ -28,7 +29,7 @@ class DataSerializer:
 
     class MarkerDepthMax:
         def __repr__(self) -> str:
-            return "..."
+            return SYMBOL_ELLIPSIS
```
Comment (noahnu) [inline src/syrupy/extensions/amber.py]: Even though we use the same symbol for context & max depth, semantically they're different and could theoretically change.

Comment (iamogbz) [reply src/syrupy/extensions/amber.py]: Sure, but then we could split the symbols apart at that point

File: src/syrupy/extensions/base.py [hunk 3]
```diff
@@ -214,36 +223,140 @@ def __ensure_snapshot_dir(self, *, index: int) -> None:
...
+                count_leading_whitespace: Callable[[str], int] = (
+                    lambda s: len(s) - len(s.lstrip())  # noqa: E731
+                )
+                if self._context_line_count:
+                    num_space = (
+                        count_leading_whitespace(lines[self._context_line_count - 1])
+                        + count_leading_whitespace(lines[-self._context_line_count])
+                    ) // 2
+                else:
+                    num_space = count_leading_whitespace(lines[num_lines // 2])
```
Comment (iamogbz) [inline src/syrupy/extensions/base.py]: Simpler to reason about but leaves room for `...` having weird indentation
```suggestion
                mid_line = lines[num_lines // 2]
                num_space = len(mid_line) - len(mid_line.lstrip())
```

File: src/syrupy/extensions/base.py [hunk 4]
```diff
@@ -214,36 +222,141 @@ def __ensure_snapshot_dir(self, *, index: int) -> None:
...
+    def __diff_data(self, a: str, b: str) -> Iterator["DiffedLine"]:
+        staged_diffed_line: Optional["DiffedLine"] = None
+        for line in ndiff(a.splitlines(keepends=True), b.splitlines(keepends=True)):
+            is_context_line = line[0] == " "
+            is_snapshot_line = line[0] == "-"
+            is_received_line = line[0] == "+"
+            is_diff_line = line[0] == "?"
+
+            if is_context_line or is_diff_line:
+                line = self.__strip_ends(line)
```
Comment (noahnu) [inline src/syrupy/extensions/base.py]: I was originally thinking of only showing the new line endings (i.e. the symbols) _if_ necessary. So if the line endings are the only change.

File: tests/test_extension_base.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+
+@pytest.mark.parametrize("Reporter", [SnapshotReporter, SnapshotReporterNoContext])
+class TestSnapshotReporter:
+    @pytest.mark.parametrize(
+        "a, b",
+        [
+            (
+                "line 0\nline 1\nline 02\nline 3\nline 4\r\nline 5\nline 6\nline 7",
+                "line 0\nline 1\nline 2\r\nline 3\nline 04\nline 5\nline 6\nline 7",
```
Comment (noahnu) [inline tests/test_extension_base.py]: more cases:
- trailing newlines
- where the difference between 2 sets is extra whitespace (space, tab) at the end
- different number of lines (additions + deletions)

**Comment by noahnu** [comment]
LGTM

**Pull request merged**